### PR TITLE
[JENKINS-59412] Clarify that build history does not include pipeline stages

### PR DESCRIPTION
**Pull request by jsoref**
See [JENKINS-59412](https://issues.jenkins-ci.org/browse/JENKINS-59412).

### Proposed changelog entries

* [JENKINS-59412](https://issues.jenkins-ci.org/browse/JENKINS-59412), Clarify that build history does not include pipeline stages

### Submitter checklist

- [x] JIRA issue is well described
- [ ] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

### CODE REVIEW

**Comment by oleg-nenashev** [comment]
I would rather say it is a bug in the Pipeline implementation. If we include Documentation for it, I would suggest creating a Pipeline ticket and referencing it from docs. https://issues.jenkins-ci.org/browse/JENKINS-38877 would do the job. I see it is "won't do", but better than nothing

**Comment by jglick** [comment]
> referencing it from docs

I think this was referring to the proposed changelog entries in the PR description?

**Review by jglick — APPROVED** [review]
While you are it, you could add

```html
<blink>While you are looking at this message,
hundreds of tiny XML files are being read, parsed, and then forgotten
as your server runs low on memory.
Navigate away to stop it.</blink>
```

:grin:

**Comment by jsoref** [comment]
@jglick: I may do that in the distant future as a distinct PR. Feel free to file a bug against me as a reminder.

**Comment by oleg-nenashev** [comment]
> I think this was referring to the proposed changelog entries in the PR description?

I meant referencing from the code. But whatever works

**Review by oleg-nenashev — APPROVED** [review]
Will merge it tomorrow if no negative feedback. It definitely improves the situation

File: core/src/main/resources/hudson/model/Computer/builds.properties [hunk 1]
```diff
@@ -20,4 +20,5 @@
 # OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 # THE SOFTWARE.
 
-title=Build History on {0}
\ No newline at end of file
+title=Build History on {0}
+disclaimer=This does not include stages from pipeline jobs.
```
Comment (KostyaSha) [inline core/src/main/resources/hudson/model/Computer/builds.properties]: does core depend on `pipeline` to hardcode additional plugin job type naming?

**Review by KostyaSha — CHANGES_REQUESTED** [review]
How core relates to pipeline job?

File: core/src/main/resources/hudson/model/Computer/builds.properties (hunk 1 above)
Comment (jsoref) [reply core/src/main/resources/hudson/model/Computer/builds.properties]: It isn't ignorant of it, including some whitelisting:
https://github.com/jenkinsci/jenkins/search?q=pipeline&unscoped_q=pipeline

If you really think that pipeline should hook these files and can provide an example of how to do it, I suppose I could look into it.

Comment (KostyaSha) [reply core/src/main/resources/hudson/model/Computer/builds.properties]: We have huge instance with custom job type and we are not using any pipeline plugins, like there are a lot of people still using freestyle jobs. Core makes confusion for users.

**Comment by jsoref** [comment]
Frankly, I could have a dozen pipeline jobs and someone could disable the pipeline plugins, leaving 0 traces of it enabled, but the jobs could still be in history. Someone could then come along and ask "what was this agent doing?" It wouldn't be unreasonable for them to come to these pages and be surprised not to see the answer they're expecting.

It's a tiny message at the top of a page. People who see what they're looking for won't dwell on it. People who don't see what they're looking for will be more likely to notice and get the answer they're looking for.

**Comment by daniel-beck** [comment]
We've referenced plugins in core for many years. It's a pragmatic thing to do; docs (and I count inline messages like this as such) can cross-reference things even when you wouldn't do it as an "architecture". So from that POV there's nothing wrong with doing so here. Examples:

https://github.com/jenkinsci/jenkins/blob/f0dadba464a56e315520b36f9047192c66be5ef6/war/src/main/webapp/help/project-config/triggerRemotely.html#L15
https://github.com/jenkinsci/jenkins/blob/22aa2e6e766074d11249893e3f35e0b99e20d3d0/war/src/main/webapp/help/project-config/downstream.html#L17...L19
https://github.com/jenkinsci/jenkins/blob/652627a2049a67c18b25024f5fbcd7d6485e4985/core/src/main/resources/hudson/security/HudsonPrivateSecurityRealm/help-allowsSignup.html#L9

Perhaps the message could be made more specific to clearly refer to a plugin, but IMO not essential -- while it's certainly possible to run "classic" Jenkins with Freestyle etc., it's unlikely active users are completely unaware of Pipeline's existence.

**Review by daniel-beck — APPROVED** [review]

**Comment by oleg-nenashev** [comment]
As a compromise, we could put something like "this history does not guarantee to include all subtasks being executed on the node, e.g. Pipeline subtasks will not be displayed

File: core/src/main/resources/hudson/model/Computer/builds.properties (hunk 1 above)
Comment (olivergondza) [reply core/src/main/resources/hudson/model/Computer/builds.properties]: @KostyaSha, would it make more sense for you if that would be a link pointing to https://jenkins.io/solutions/pipeline/ ?

**Review by olivergondza — APPROVED** [review]

**Comment by oleg-nenashev** [comment]
@KostyaSha would you agree with the amendment suggested above?

**Comment by KostyaSha** [comment]
> @KostyaSha would you agree with the amendment suggested above?

well, at least.

**Comment by oleg-nenashev** [comment]
Somebody still needs to amend it :(

File: core/src/main/resources/hudson/model/Computer/builds.properties (hunk 1 above)
Comment (oleg-nenashev) [reply core/src/main/resources/hudson/model/Computer/builds.properties]:
```suggestion
disclaimer=This history does not guarantee to include all subtasks being executed on the node, e.g. Jenkins Pipeline subtasks will not be displayed.
```

File: core/src/main/resources/hudson/model/View/builds.properties [hunk 2]
```diff
@@ -20,4 +20,5 @@
 # OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 # THE SOFTWARE.
 
-buildHistory=Build History of {0}
\ No newline at end of file
+buildHistory=Build History of {0}
+disclaimer=This does not include stages from pipeline jobs.
```
Comment (oleg-nenashev) [inline core/src/main/resources/hudson/model/View/builds.properties]:
```suggestion
disclaimer=This history does not guarantee to include all subtasks being executed on the node, e.g. Jenkins Pipeline subtasks will not be displayed.
```

**Review by oleg-nenashev — CHANGES_REQUESTED** [review]
Requesting the change discussed above so that we can finally land it

File: core/src/main/resources/hudson/model/Computer/builds.properties (hunk 1 above)
Comment (jsoref) [reply core/src/main/resources/hudson/model/Computer/builds.properties]: `does not guarantee to` -> `is not guaranteed to`

The former is wrong for some reason I can't quite explain.

**Comment by varyvol** [comment]
@jsoref this seems to be ready except for @oleg-nenashev 's suggestions. Could you evaluate them and, if you think they are fine, accept them so this can be merged?

File: core/src/main/resources/hudson/model/Computer/builds.properties (hunk 1 above)
Comment (jsoref) [inline core/src/main/resources/hudson/model/Computer/builds.properties]:
```suggestion
disclaimer=This history is not guaranteed to include all subtasks executed on the node, e.g. Jenkins Pipeline subtasks will not be displayed.
```

File: core/src/main/resources/hudson/model/View/builds.properties (hunk 2 above)
Comment (jsoref) [reply core/src/main/resources/hudson/model/View/builds.properties]:
```suggestion
disclaimer=This history is not guaranteed to include all subtasks executed on the node, e.g. Jenkins Pipeline subtasks will not be displayed.
```

**Review by oleg-nenashev — APPROVED** [review]
I plan to merge it tomorrow if no negative feedback

**Review by fcojfernandez — APPROVED** [review]

File: core/src/main/resources/hudson/model/User/builds.properties [hunk 3]
```diff
@@ -20,3 +20,4 @@
 # OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 # THE SOFTWARE.
 title=Builds for {0}
+disclaimer=This does not include stages from pipeline jobs.
```
Comment (jglick) [inline core/src/main/resources/hudson/model/User/builds.properties]: Why the different wording than the other ones?

Would be nice to define this message in only one place.

File: core/src/main/resources/hudson/model/Computer/builds.properties [hunk 4]
```diff
@@ -20,4 +20,5 @@
 # OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 # THE SOFTWARE.
 
-title=Build History on {0}
\ No newline at end of file
+title=Build History on {0}
+disclaimer=This history is not guaranteed to include all subtasks executed on the node, e.g. Jenkins Pipeline subtasks will not be displayed.
```
Comment (jglick) [inline core/src/main/resources/hudson/model/Computer/builds.properties]: Are users going to know what “subtasks” are? When you write the script, you are using the `node` step, and I do not think the expression “subtask” is used anywhere in the GUI.

**Comment by oleg-nenashev** [comment]
@jglick You did not remove the `ready-for-merge` label. Do you consider your comments as blocking ones?

**Comment by jglick** [comment]
> Do you consider your comments as blocking ones?

No.

**Comment by oleg-nenashev** [comment]
Let's get it over the line finally

**Pull request merged**